### ISSUE TITLE
Recursive type aliases don't handle Lists and Dicts correctly

### ISSUE THREAD

**Issue by judej** [microsoft/pyright#1021]
When using recursive type aliases inside a List of Dict, the type checker inaccurately flags usage as error

**To Reproduce**
```
Foo = Union[bool, List["Foo"], Dict[str, "Foo"]]

# No error
bar1: Foo = [True]

# Expression of type "List[bool | List[bool]]" cannot be assigned to declared type "Foo"
#   Type "List[bool | List[bool]]" cannot be assigned to type "Foo"
#     "List[bool | List[bool]]" is incompatible with "bool"
#     "List[bool | List[bool]]" is incompatible with "Dict[str, Foo]"
#       TypeVar "_T" is invariant
#         Type "bool | List[bool]" cannot be assigned to type "Foo"Pyright 
bar: Foo = [True, [True, False]]
```

**VS Code extension or command-line**
VSCode Extension

**Additional context**
PyRight 1.1.69

**Comment by erictraut** [comment]
Thanks for finding and reporting this issue. It was a problem with an interaction between recursive type aliases and bidirectional type inference of list and dict statements. I've introduced a fix, and it will be in the next release.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.70, which I just published. It will also be in the next version of Pylance.

**Comment by tristan957** [comment]
@erictraut I seem to still be experiencing this issue as of VSCode extension Pylance release 2020.10.0 which released today I believe.

Given:

```python
Config = Dict[
    str,
    Optional[
        Union[str, int, float, bool, List[Optional[Union[str, float, int]]], "Config"]
    ],
]

x: Config = {"fsd": {"fsd": 1, "fsdf": 1}}
```

Pylance reports 
```
Expression of type "Dict[str, Dict[str, int]]" cannot be assigned to declared type "Config"
  TypeVar "_VT" is invariant
    Type "Dict[str, int]" cannot be assigned to type "str | int | float | bool | List[str | float | int | None] | Config | None"
```

The error message makes sense only because the type it is trying to assign the dict to is not the actual type of `Config`. It took out the `Dict[str. ...]` part of the type definition.

**Comment by erictraut** [comment]
@tristan957, I'm not able to repro the problem with the code snippet above. Can you double check that there's not something missing from your example? I've tried with the latest version of Pylance and Pyright. I've tried enabling strict type checking even, and I don't receive any errors with the above code.

![Screen Shot 2020-10-08 at 1 37 04 PM](https://user-images.githubusercontent.com/7040122/95518239-8611be00-0977-11eb-8897-1edbb01a703a.png)

**Comment by tristan957** [comment]
Eric, I was being a complete buffoon somehow. Didn't delete my comment before the github notification went out. I think the language server didn't pick up my updated pyi file, so a server restart did the trick.

**Comment by erictraut** [comment]
OK, glad to hear it's working for you now!
